server/src/main/java/io/druid/client/DirectDruidClient.java

```diff
@@ -257,7 +264,9 @@ public int getNumOpenConnections()
                   )
               );
             }
-            queue.put(new ChannelBufferInputStream(response.getContent()));
+
+            checkBufferCapacity(bytes, queryBufferingTimeout);
```

Comment:
I think this is going to be a busy-wait-loop in a netty thread (IIRC that is what `handleResponse` gets called in) so this could cause starvation of the netty thread pool.

Am I reading this right? If so is there some better way we can do backpressure with netty?

I'm not a netty expert so I'm not sure what the best way is to go here. Ideally we want netty to stop reading from the socket and stop delivering content to us, which would propagate to the historical nodes, which would eventually block up their http threads. But I don't know how to make that happen with netty.

Reply:
http://normanmaurer.me/presentations/2014-facebook-eng-netty/slides.html#33.0 seems maybe relevant.

Reply:
@gianm True. It might cause starvation and that's why I introduced `queryBufferingTimeout`. This is a workaround to fix the OOMs issue with the broker, I am not a netty expert neither, not sure what will be the best to do backpressure with netty.

Reply:
Are you interested in exploring a way to do this without blocking the netty threads? That would really be an ideal solution I think.

Reply:
Seems like netty also supports some watermark buffers which can possibly be used to achive backpressure. 
Atleast the javadocs here seems to indicate this - 
http://netty.io/4.1/api/io/netty/channel/WriteBufferWaterMark.html
However this might need an upgrade to netty version used by druid.

Reply:
@gianm I would love to explore what I can do without blocking netty threads.